sshd-core/src/test/java/org/apache/sshd/common/file/root/RootedFileSystemProviderTest.java

```diff
@@ -0,0 +1,270 @@
+package org.apache.sshd.common.file.root;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Random;
+import java.util.TreeSet;
+import org.junit.BeforeClass;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+
+
+/**
+ * Tests the RootedFileSystemProvider implementation of {@link java.nio.file.spi.FileSystemProvider}
+ * checking that permissions for generic FS commands are not permitted outside of the
+ * root directory.
+ *
+ * Individual tests are form pairs (e.g. testX, testXInvalid) where testXInvalid is
+ * expected to test a parent path of {@link RootedFileSystem#getRoot()}
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class RootedFileSystemProviderTest extends AssertableFile {
+
+  private static RootedFileSystem fileSystem;
+  private static Path rootSandbox;
+
+  public RootedFileSystemProviderTest() {
+    super();
+  }
+
+  @BeforeClass
+  public static void onlyOnce() throws IOException {
+    try {
+      rootSandbox = FileHelper.createTestSandbox();
+      fileSystem = (RootedFileSystem) new RootedFileSystemProvider()
+          .newFileSystem(rootSandbox, Collections.EMPTY_MAP);
+    } catch (IOException e) {
+      fail("Could not create test sandbox: " + e.getMessage());
+    }
+  }
+
+  @Test
+  public void testRoot() {
+    assertTrue(isFile(fileSystem.getRoot()) &&
+        isDir(fileSystem.getRoot()) &&
+        isReadable(fileSystem.getRoot()) &&
+        isRootedAt(rootSandbox, fileSystem.getRoot()));
+  }
+
+  /* mkdir */
+  @Test
+  public void testMkdir() throws IOException {
+    Path created = FileHelper.createDirectory(fileSystem.getPath(getCurrentTestName()));
+    assertTrue(isFile(created) && isDir(created) && isReadable(created));
+  }
+
+  @Test(expected = InvalidPathException.class)
+  public void testMkdirInvalid() throws IOException {
+    FileHelper.createDirectory(fileSystem.getPath("../" + getCurrentTestName()));
+  }
+
```

Comment:
Even though the _expected_ attribute takes care of this, I usually recommend adding a specific _fail_ call that gives more information so the failure can be more easily diagnosed:
```java
@Test(expected=...)
public void testExpectedFailure() {
    ....
    result = callThatShouldHaveFailed(...);
    fail("Unexpected success: result=" + result);
}
```